src/Client/ClientBase.cpp

```diff
@@ -0,0 +1,1367 @@
+        else
+            block_out_stream = global_context->getOutputStream(current_format, out_file_buf ? *out_file_buf : *out_buf, block);
+
+        block_out_stream->writePrefix();
+    }
+}
+
+
+void ClientBase::initLogsOutputStream()
+{
+    if (!logs_out_stream)
+    {
+        WriteBuffer * wb = out_logs_buf.get();
+
+        if (!out_logs_buf)
+        {
+            if (server_logs_file.empty())
+            {
+                /// Use stderr by default
+                out_logs_buf = std::make_unique<WriteBufferFromFileDescriptor>(STDERR_FILENO);
+                wb = out_logs_buf.get();
+            }
+            else if (server_logs_file == "-")
+            {
+                /// Use stdout if --server_logs_file=- specified
+                wb = &std_out;
+            }
+            else
+            {
+                out_logs_buf
+                    = std::make_unique<WriteBufferFromFile>(server_logs_file, DBMS_DEFAULT_BUFFER_SIZE, O_WRONLY | O_APPEND | O_CREAT);
+                wb = out_logs_buf.get();
+            }
+        }
+
+        logs_out_stream = std::make_shared<InternalTextLogsRowOutputStream>(*wb, stdout_is_a_tty);
+        logs_out_stream->writePrefix();
+    }
+}
+
+
+void ClientBase::processTextAsSingleQuery(const String & full_query)
+{
+    /// Some parts of a query (result output and formatting) are executed
+    /// client-side. Thus we need to parse the query.
+    const char * begin = full_query.data();
+    auto parsed_query = parseQuery(begin, begin + full_query.size(), false);
+
+    if (!parsed_query)
+        return;
+
+    String query_to_execute;
+
+    // An INSERT query may have the data that follow query text. Remove the
+    /// Send part of query without data, because data will be sent separately.
+    auto * insert = parsed_query->as<ASTInsertQuery>();
+    if (insert && insert->data)
+        query_to_execute = full_query.substr(0, insert->data - full_query.data());
+    else
+        query_to_execute = full_query;
+
+    processParsedSingleQuery(full_query, query_to_execute, parsed_query);
+
+    if (have_error)
+        processError(full_query);
+}
+
+
+void ClientBase::processOrdinaryQuery(const String & query_to_execute, ASTPtr parsed_query)
+{
+    /// Rewrite query only when we have query parameters.
+    /// Note that if query is rewritten, comments in query are lost.
+    /// But the user often wants to see comments in server logs, query log, processlist, etc.
+    auto query = query_to_execute;
+    if (!query_parameters.empty())
+    {
+        /// Replace ASTQueryParameter with ASTLiteral for prepared statements.
+        ReplaceQueryParameterVisitor visitor(query_parameters);
+        visitor.visit(parsed_query);
+
+        /// Get new query after substitutions. Note that it cannot be done for INSERT query with embedded data.
+        query = serializeAST(*parsed_query);
+    }
+
+    int retries_left = 10;
+    for (;;)
+    {
+        assert(retries_left > 0);
```

Comment:
> Why do we need automatic retries

There is a comment for this a little below inside catch.

> why do we need such assertion?

Took it from previous client.cpp, but now refactored this loop slightly better.
Looks like it is needed because we want to believe we won't get "client should retry" from server more than a few times in a row for the same query.

Reply:
> There is a comment for this a little below inside catch.

Yes, I saw the comment about "Client should retry" errors, I'm just not sure that we really need automatic retries in this case.

> Took it from previous client.cpp

I know, but now you are in git blame, so everyone will ask you =)